### PR TITLE
[CELEBORN-568] Support storage type selection

### PR DESCRIPTION
**Pull request by FMX**
### What changes were proposed in this pull request?
1. Celeborn supports storage type selection. HDD, SSD, and HDFS are available for now.
2. Add new buffer size for HDFS file writers.
3. Worker support empty working dirs.

### Why are the changes needed?
Support HDFS only scenario.

### Does this PR introduce _any_ user-facing change?
NO.

### How was this patch tested?
UT and cluster.

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -2116,6 +2138,14 @@ object CelebornConf extends Logging {
       .bytesConf(ByteUnit.BYTE)
       .createWithDefaultString("256k")
 
+  val WORKER_HDFS_FLUSHER_BUFFER_SIZE: ConfigEntry[Long] =
+    buildConf("celeborn.worker.flusher.hdfs.buffer.size")
+      .categories("worker")
+      .version("0.3.0")
+      .doc("Size of buffer used by a HFDS flusher.")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2145]:
```suggestion
      .doc("Size of buffer used by a HDFS flusher.")
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -2077,7 +2099,7 @@ object CelebornConf extends Logging {
     buildConf("celeborn.worker.commitFiles.timeout")
       .withAlternative("celeborn.worker.shuffle.commit.timeout")
       .categories("worker")
-      .doc("Timeout for a Celeborn worker to commit files of a shuffle.")
+      .doc("Timeout for a Celeborn worker to commit files of a shuffle. If celeborn is writing to HDFS, increase this value to 240s or larger.")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2102]:
```suggestion
      .doc("Timeout for a Celeborn worker to commit files of a shuffle. It's recommended to set at least `240s` when `HDFS` is enabled in `celeborn.storage.activeLevels`.")
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -3546,4 +3576,13 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(true)
 
+  val ACTIVE_STORAGE_LEVELS: ConfigEntry[String] =
+    buildConf("celeborn.storage.activeLevels")
+      .categories("master", "worker")
+      .version("0.3.0")
+      .doc("Enabled storage levels. Optional values: HDD,SSD,HDFS. ")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:3583]: question: how do you treat S3, OSS? independent options, or alias of HDFS?

Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:3583]: nit: "Optional values" => "available options", keep consistent with others

Comment (FMX) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:3588]: S3, OSS and other storage levels will be added when ready.

File: common/src/main/java/org/apache/celeborn/common/protocol/StorageInfo.java [hunk 4]
```diff
@@ -49,6 +49,12 @@ public int getValue() {
           }
         }
       };
+  public static Set<String> typeNames =
+      new HashSet() {
```
Comment (pan3793) [inline common/src/main/java/org/apache/celeborn/common/protocol/StorageInfo.java:53]: this actually creates a subclass, and not much readable.

suggest:
```
  public static Set<String> typeNames;
  public static Map<Integer, Type> typesMap;

  static {
    typeNames = new HashSet<>();
    typesMap = new HashMap<>();
    for (Type type : Type.values()) {
      typeNames.add(type.name());
      typesMap.put(type.value, type);
    }
  }
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 5]
```diff
@@ -497,6 +497,21 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
...
+  def activeStorageLevels: JHashSet[Type] = {
+    val set = new JHashSet[StorageInfo.Type]()
+    get(ACTIVE_STORAGE_LEVELS).split(",").foreach(f => {
+      if (StorageInfo.validateStorageType(f)) {
+        set.add(StorageInfo.Type.valueOf(f))
+      } else {
+        logWarning(s"Found invalid storage level ${f}")
+      }
+    })
+    set
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:510]:
```
    val (valid, invalid) = str.split(",").partition(StorageInfo.validateStorageType)
    logWarning(s"Ignore invalid storage levels: $invalid")
    valid
```

File: worker/pom.xml [hunk 6]
```diff
@@ -76,7 +76,6 @@
       <groupId>org.apache.logging.log4j</groupId>
       <artifactId>log4j-1.2-api</artifactId>
     </dependency>
-
```
Comment (pan3793) [inline worker/pom.xml:79]: it divide different scopes, is fine.

File: common/src/test/scala/org/apache/celeborn/common/CelebornConfSuite.scala [hunk 7]
```diff
@@ -186,4 +186,25 @@ class CelebornConfSuite extends CelebornFunSuite {
...
+    assert(conf.workerBaseDirs.isEmpty)
+
+    conf.set("celeborn.storage.activeLevels", "SDD,HDD")
+    assert(!conf.workerBaseDirs.isEmpty)
+  }
+
+  test("Test commit file threads") {
+    val conf = new CelebornConf()
+    conf.set("celeborn.storage.activeLevels", "HDFS")
+    assert(256 == conf.workerCommitThreads)
```
Comment (pan3793) [inline common/src/test/scala/org/apache/celeborn/common/CelebornConfSuite.scala:205]: The right pattern for scalatest assert is `assert(actual === expected)`

```suggestion
    assert(conf.workerCommitThreads === 256)
```

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java [hunk 8]
```diff
@@ -202,7 +202,8 @@ public void updateWorkerHeartbeatMeta(
           });
     }
     appDiskUsageMetric.update(estimatedAppDiskUsage);
-    if (!blacklist.contains(worker) && disks.isEmpty()) {
+    // If using HDFSONLY mode, workers with empty disks should not be put into blacklist.
```
Comment (pan3793) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:205]:
```suggestion
    // If using HDFS_ONLY mode, workers with empty disks should not be put into blacklist.
```

File: docs/configuration/master.md [hunk 9]
```diff
@@ -33,4 +33,5 @@ license: |
 | celeborn.master.slot.assign.loadAware.numDiskGroups | 5 | This configuration is a guidance for load-aware slot allocation algorithm. This value is control how many disk groups will be created. | 0.3.0 | 
 | celeborn.master.slot.assign.policy | ROUNDROBIN | Policy for master to assign slots, Celeborn supports two types of policy: roundrobin and loadaware. | 0.3.0 |
```
Comment (pan3793) [inline docs/configuration/master.md:34]: The description should updated?

**Comment by FMX** [comment]
> `Support HDFS only scenario.` I don't quite understand `HDFS Only`, seems local disk and hdfs can be used together.

Celeborn won't work if the Worker doesn't have local disks. So this PR will allow users to use Celeborn without local disks. That's what I mean.

**Comment by waitinfuture** [comment]
> 

I see, better to doc about the supported configurations/combinations.

**Comment by FMX** [comment]
> Member

I think this section could be added to the deployment docs.

File: docs/deploy.md [hunk 10]
```diff
@@ -68,8 +82,22 @@ celeborn.master.ha.ratis.raft.server.storage.dir /mnt/disk1/rss_ratis/
 celeborn.metrics.enabled true
 # If you want to use HDFS as shuffle storage, make sure that flush buffer size is at least 4MB or larger.
 celeborn.worker.flusher.buffer.size 256k
-# Disk type is HDD by default.
+
+# If Celeborn workers have local disks, use following config.
+# Disk type is HDD by defaut.
```
Comment (pan3793) [inline docs/deploy.md:87]:
```suggestion
# Disk type is HDD by default.
```

File: docs/deploy.md [hunk 11]
```diff
@@ -137,12 +165,16 @@ spark.celeborn.client.spark.shuffle.writer hash
...
 # If you have only one worker, this setting must be false 
+# If your Celeborn is using HDFS, it's recommended to set this setting to false
 spark.celeborn.client.push.replicate.enabled true
 
 # Support for Spark AQE only tested under Spark 3
 # we recommend set localShuffleReader to false to get better performance of Celeborn
 spark.sql.adaptive.localShuffleReader.enabled false
 
+# If Celeborn is using HDFS
+spark.celeborn.worker.storage.hdfs.dir hdfs://[your-hdfs-namespace]/celeborn
```
Comment (pan3793) [inline docs/deploy.md:176]:
```suggestion
spark.celeborn.worker.storage.hdfs.dir hdfs://<namenode->/celeborn
```

**Review by pan3793 — APPROVED** [review]

**Comment by waitinfuture** [comment]
> 

Thanks, please also test and doc the scenario where both local disks and hdfs exist.

**Comment by FMX** [comment]
Here are some performance test results. I test this patch on a tiny cluster with 4 nodes (8 vCPU 32GB).
Depending on the partition num, Celeborn on HDFS will be slower by up to 50%.
Please don't use Celeborn on HDFS if your HDFS cluster is small.

截屏2023-06-27 17 42 38

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks! Merging to main/0.3

**Pull request closed without merging**